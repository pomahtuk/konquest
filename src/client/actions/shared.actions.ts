export enum SharedActionTypes {
  RESTORE_GAME = "RESTORE_GAME"
}

export interface RestoreGameAction {
  type: typeof SharedActionTypes.RESTORE_GAME;
  storedGame?: string;
}

export function restoreGame(storedGame?: string): RestoreGameAction {
  return { type: SharedActionTypes.RESTORE_GAME, storedGame };
}
