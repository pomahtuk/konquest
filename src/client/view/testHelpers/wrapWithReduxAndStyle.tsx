import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
