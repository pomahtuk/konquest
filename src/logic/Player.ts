export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
