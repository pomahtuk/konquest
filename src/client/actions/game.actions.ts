import Player, { PlayerTurn } from "../../logic/Player";

/*
 * action types
 */
export enum GameActionTypes {
  START_GAME = "START_GAME",
  SET_GAME_OPTIONS = "SET_GAME_OPTIONS",
  ADD_PLAYER_TURN = "ADD_PLAYER_TURN"
}

export interface StateGameOptions {
  fieldSize: number;
  neutralPlanetCount: number;
  players: Player[];
}

export interface StartGameAction {
  type: typeof GameActionTypes.START_GAME;
}

export interface SetGameOptionsAction {
  type: typeof GameActionTypes.SET_GAME_OPTIONS;
  gameOptions: StateGameOptions;
}

export interface AddPlayerTurnAction {
  type: typeof GameActionTypes.ADD_PLAYER_TURN;
  turnData: PlayerTurn;
}
/*
 * action creators
 */
export function startGame(): StartGameAction {
  return { type: GameActionTypes.START_GAME };
}

export function setGameOptions(gameOptions: StateGameOptions): SetGameOptionsAction {
  return { type: GameActionTypes.SET_GAME_OPTIONS, gameOptions };
}

export function addPlayerTurn(turnData: PlayerTurn): AddPlayerTurnAction {
  return { type: GameActionTypes.ADD_PLAYER_TURN, turnData };
}
