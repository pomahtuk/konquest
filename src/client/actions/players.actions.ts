import Player from "../../logic/Player";

export enum PlayerActionTypes {
  ADD_PLAYER = "ADD_PLAYER",
  REMOVE_PLAYER = "REMOVE_PLAYER",
  CLEAR_PLAYERS = "CLEAR_PLAYERS"
}

export interface AddPlayerAction {
  type: typeof PlayerActionTypes.ADD_PLAYER;
  player: Player;
}

export interface RemovePlayerAction {
  type: typeof PlayerActionTypes.REMOVE_PLAYER;
  player: Player;
}

export interface ClearPlayersAction {
  type: typeof PlayerActionTypes.CLEAR_PLAYERS;
}

// action creators
export function addPlayer(player: Player): AddPlayerAction {
  return { type: PlayerActionTypes.ADD_PLAYER, player };
}

export function removePlayer(player: Player): RemovePlayerAction {
  return { type: PlayerActionTypes.REMOVE_PLAYER, player };
}

export function clearPlayers(): ClearPlayersAction {
  return { type: PlayerActionTypes.CLEAR_PLAYERS };
}
