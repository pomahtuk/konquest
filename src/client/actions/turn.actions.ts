import { PlayerTurnOrder } from "../../logic/Player";
import Planet from "../../logic/Planet";

/*
 * action types
 */
export enum TurnActionTypes {
  SET_ORIGIN_PLANET = "SET_ORIGIN_PLANET",
  SET_DESTINATION_PLANET = "SET_DESTINATION_PLANET",
  ADD_PLAYER_TURN_ORDER = "ADD_PLAYER_TURN_ORDER",
  REMOVE_PLAYER_TURN_ORDER = "REMOVE_PLAYER_TURN_ORDER",
  CLEAR_TURN_DATA = "CLEAR_TURN_DATA",
  UNSET_PLANETS = "UNSET_PLANETS"
}

export interface SetPlanetAction {
  type: typeof TurnActionTypes.SET_ORIGIN_PLANET | typeof TurnActionTypes.SET_DESTINATION_PLANET;
  planet?: Planet;
}

export interface AddPlayerTurnOrderAction {
  type: typeof TurnActionTypes.ADD_PLAYER_TURN_ORDER;
  order: PlayerTurnOrder;
}

export interface ClearTurnDataAction {
  type: typeof TurnActionTypes.CLEAR_TURN_DATA;
}

export interface RemovePlayerTurnOrderAction {
  type: typeof TurnActionTypes.REMOVE_PLAYER_TURN_ORDER;
  index: number;
}

export interface UnsetPlanetsAction {
  type: typeof TurnActionTypes.UNSET_PLANETS;
}

/*
 * action creators
 */
export function setOriginPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_ORIGIN_PLANET, planet };
}

export function setDestinationPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_DESTINATION_PLANET, planet };
}

export function addPlayerTurnOrder(order: PlayerTurnOrder): AddPlayerTurnOrderAction {
  return { type: TurnActionTypes.ADD_PLAYER_TURN_ORDER, order };
}

export function removePlayerTurnOrder(index: number): RemovePlayerTurnOrderAction {
  return { type: TurnActionTypes.REMOVE_PLAYER_TURN_ORDER, index };
}

export function clearTurnData(): ClearTurnDataAction {
  return { type: TurnActionTypes.CLEAR_TURN_DATA };
}

export function unsetPlanets(): UnsetPlanetsAction {
  return { type: TurnActionTypes.UNSET_PLANETS };
}
