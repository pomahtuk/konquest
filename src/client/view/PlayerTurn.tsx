import React, { ReactElement, useEffect } from "react";
import { useSelector, shallowEqual, useDispatch } from "react-redux";

import { addPlayerTurn, setDestinationPlanet, setOriginPlanet, addPlayerTurnOrder } from "../actions/game.actions";
import playerTurnSelector, { PlayerTurnStoreSlice } from "../selectors/playerTurn.selector";

import AddPlayerTurn from "./AddPlayerTurn";
import OrderList from "./OrderList";
import Button from "./foundations/Button";
import useSlider from "../hooks/useSlider";

const PlayerTurn = (): ReactElement => {
  const dispatch = useDispatch();
  const { orders, activePlayer, originPlanet, destinationPlanet }: PlayerTurnStoreSlice = useSelector(playerTurnSelector, shallowEqual);
  const { value, onChange, setValue } = useSlider(0);

  useEffect(() => {
    if (originPlanet) {
      setValue(originPlanet.ships);
    }
  }, [setValue, originPlanet]);

  const cleanUp = (): void => {
    setValue(0);
    dispatch(setOriginPlanet(undefined));
    dispatch(setDestinationPlanet(undefined));
  };

  const onAddOrder = (): void => {
    originPlanet &&
      destinationPlanet &&
      dispatch(
        addPlayerTurnOrder({
          origin: originPlanet.name,
          destination: destinationPlanet.name,
          amount: value as number
        })
      );
    cleanUp();
  };

  const onCompleteTurn = (): void => {
    dispatch(
      addPlayerTurn({
        player: activePlayer,
        orders
      })
    );
    // clean up planet selection
    cleanUp();
  };

  return (
    <div>
      <AddPlayerTurn
        activePlayer={activePlayer}
        amount={value as number}
        originPlanet={originPlanet}
        destinationPlanet={destinationPlanet}
        onOrderAmountChange={onChange}
        onAddOrder={onAddOrder}
        onCancel={cleanUp}
      />

      <OrderList orders={orders} />

      <Button variant="primary" onClick={onCompleteTurn}>
        Complete turn
      </Button>
    </div>
  );
};

export default PlayerTurn;
