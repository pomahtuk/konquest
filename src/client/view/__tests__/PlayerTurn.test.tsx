import React from "react";
import { render, fireEvent } from "@testing-library/react";

import wrapWithReduxAndStyle, { CurrentStore } from "../testHelpers/wrapWithReduxAndStyle";
import PlayerTurn from "../PlayerTurn";
import { startGame, setGameOptions } from "../../actions/game.actions";
import { setOriginPlanet, setDestinationPlanet } from "../../actions/turn.actions";
import Player from "../../../logic/Player";

const selectPlanets = (): void => {
  // get planets from store
  const {
    game: { planets }
  } = CurrentStore.getState();
  // dispatch select events
  CurrentStore.dispatch(setOriginPlanet(planets["A"]));
  CurrentStore.dispatch(setDestinationPlanet(planets["C"]));
};

describe("<PlayerTurn />", (): void => {
  beforeAll((): void => {
    // prepare game
    CurrentStore.dispatch(
      setGameOptions({
        fieldSize: 6,
        neutralPlanetCount: 1,
        players: [new Player("1"), new Player("2")]
      })
    );
    // start game
    CurrentStore.dispatch(startGame());
  });

  it("renders <PlayerTurn /> component at the beginning of player turn", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<PlayerTurn />));
    expect(container).toMatchSnapshot();
  });

  it("<PlayerTurn /> capable of basic turn modifications: add order, cancel order, complete turn", (): void => {
    const {
      game: { activePlayer: startingPlayer }
    } = CurrentStore.getState();
    const { getByText } = render(wrapWithReduxAndStyle(<PlayerTurn />));
    // now let's set selected planets
    selectPlanets();
    // at this point we should be able to add order
    const addOrderButton = getByText("Add order");
    fireEvent.click(addOrderButton);
    const {
      turn: { currentPlayerOrders }
    } = CurrentStore.getState();
    expect(currentPlayerOrders).toHaveLength(1);
    // now we let's re-select planets
    selectPlanets();
    // and try de-selecting them
    const cancelOrderButton = getByText("Cancel order");
    fireEvent.click(cancelOrderButton);
    const {
      turn: { originPlanet, destinationPlanet }
    } = CurrentStore.getState();
    expect(originPlanet).not.toBeDefined();
    expect(destinationPlanet).not.toBeDefined();
    // now let's complete turn
    const completeTurnButton = getByText("Complete turn");
    fireEvent.click(completeTurnButton);
    // at this point game should be waiting for new player
    const {
      game: { activePlayer: currentPlayer }
    } = CurrentStore.getState();
    expect(startingPlayer).not.toMatchObject(currentPlayer);
  });

  it("<PlayerTurn /> capable of removing already added turn order", (): void => {
    const { getByText, getByTestId } = render(wrapWithReduxAndStyle(<PlayerTurn />));
    // now let's set selected planets
    selectPlanets();
    // at this point we should be able to add order
    const addOrderButton = getByText("Add order");
    fireEvent.click(addOrderButton);

    // order added, should be able to remove
    const removeButton = getByTestId("remove");
    expect(removeButton).toBeDefined();
    fireEvent.click(removeButton);

    const {
      turn: { currentPlayerOrders }
    } = CurrentStore.getState();
    expect(currentPlayerOrders).toHaveLength(0);
  });
});
