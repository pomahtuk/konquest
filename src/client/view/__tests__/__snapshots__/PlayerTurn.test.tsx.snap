// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerTurn /> renders <PlayerTurn /> component at the beginning of player turn 1`] = `
.emotion-5 {
  position: relative;
  display: inline-block;
  top: 0;
  left: 0;
}

.emotion-0 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 0 0 1px;
  top: -1px;
  left: -1px;
}

.emotion-1 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 1px 0 0;
  top: -1px;
  right: -1px;
}

.emotion-2 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 0 1px 1px;
  bottom: -1px;
  left: -1px;
}

.emotion-3 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 1px 1px 0;
  bottom: -1px;
  right: -1px;
}

.emotion-4 {
  border: 1px solid #acf9fb;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  vertical-align: middle;
  position: relative;
  background-color: rgba(6,43,45,0.65);
  color: #acf9fb;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  font-family: Arial,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  z-index: 2;
}

.emotion-4:focus {
  outline: none;
}

<div>
  <div>
    <h2>
      1
       turn
    </h2>
    <div>
      Please select origin planet
    </div>
  </div>
  <div
    class="emotion-5"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
    <button
      class="emotion-4"
      type="button"
    >
      Complete turn
    </button>
  </div>
</div>
`;
